linter.py

```diff
@@ -16,3 +12,10 @@ class Phpmd(Linter):
         'selector': 'embedding.php, source.php',
         '@rulesets:,': 'cleancode,codesize,controversial,design,naming,unusedcode'
     }
+
+    def cmd(self):
+        if self.settings['lint_mode'] != 'background':
+            self.tempfile_suffix = '-'
+            return ('phpmd', '${file_on_disk}', 'text')
+
+        return ('phpmd', '${temp_file}', 'text')
```

Comment:
Of course this would not make the "manual" trigger safe to use.   That's a bit the question:  should we disallow a manual trigger because it would somehow result in garbage.

Reply:
Indeed, I did not consider the state of the view. I was approaching as if I would set the required behaviour globally or in the project, and then start using the linter. What you describe would indeed be my concern, that a manual trigger would return different results from any other live mode. For instance, I sometimes just hit "lint this view" to make sure the linters have run, and that should be consistent with e.g. just saving the file etc.

I have a bit of a struggle getting back into the mindset of all the different modes and interactions SL supports .. it being Friday and all 😅 But my initial reaction would be that your proposal is overly subtle 🤔

Reply:
> I have a bit of a struggle getting back into the mindset of all the different modes and interactions SL supports

Well, I actually cheat and have to read the source. And the documentation about capturing filename I wrote yesterday.  

So your approach: the user sets "manual", hits "lint this view" on dirty buffers but it actually still lints the file on disk.  All this even when the users doesn't use any baseline stuff.  

The subtle approach:  the users uses baseline stuff, and for example "load_save" mode, and then uses "lint this view" on an unsaved file, then SL will switch to temporary filenames and phpmd will produce noise not respecting the baseline.  The user could now roll eyes and fix it by saving and linting again.  

The latter is slightly better as a) the user can roll-eyes and fix the situation and b) the other user not using baselines is not affected at all.  Am I correct?

All can probably be solved with an extra setting `only_lint_saved_files` and code such as:

```python
    @classmethod
    def should_lint(cls, view, settings, reason):
        # type: (sublime.View, LinterSettings, Reason) -> bool
        """Decide whether the linter can run at this point in time."""
        only_lint_saved_files = settings["only_lint_saved_files"]
        if only_lint_saved_files and (
            view.is_dirty() or not view.file_name()
        ):
            return False

        return super().should_lint(cls, view, settings, reason)

    def cmd(self):
        only_lint_saved_files = self.settings["only_lint_saved_files"]
        target = '$file_on_disk' if only_lint_saved_files else '$temp_file'
        return ('phpmd', target, 'text')
```

Reply:
Fair points. Not sure there is a black-and-white good answer here. But always good to discuss.

> the user sets "manual", hits "lint this view" on dirty buffers but it actually still lints the file on disk

This is how real-file-only linters always work right?

>  All this even when the users doesn't use any baseline stuff.

Presumably most users will never change the lint mode anyway, unless (in this case) consciously doing so to support baseline stuff. That's the assumption that allows us to change the implementation based on the lint mode setting in the first place. Alternatives might have fewer assumptions, but come with their own downsides.

> The user could now roll eyes and fix it by saving and linting again.

This also holds the assumption that that's not disorienting, frustrating or confusing. I work on files that have hundreds of violations without the baseline. It would be very weird (and feel broken) if in one mode it reports 150 errors and in another it reports 0. I would argue it's easier to understand that you need to save the file for the linter to see your changes. That's how the entire load_save and save modes operate anyway: the error markers are potentially outdated until you save.

Reply:
> This is how real-file-only linters always work right?

No, real-file linters only run on saved files.  If you hit "lint this view" they (silently) skip. And they must do so because when we run  `('phpmd', '${file_on_disk}', 'text')` on a dirty buffer, we lint the file on disk and would highlight the wrong regions because the view is dirty,  That is really a bug imo. 

> Presumably most users will never change the lint mode anyway

Users already run in "manual" mode and we can't change the semantics of it, that's the problem.  It would be a breaking change, and we basically loose the "lint this view" at your will functionality.  Which is not while-you-type but whenever-you-want.  

Eye-rolling means an annoyance with a work around.  For all linters (at least typically) running on stdin or a temp_file when dirty or file_on_disk when just saved should yield the exact same output.   So here it does not, and if that's too confusing/eye-rolling we maybe should add a switch "only run on saved files".

Reply:
Dang, so it essentially comes down to if we forbid the manual "lint this view" *trigger*.  Because the same is true for load and save mode, we had to skip for a dirty buffers to not get garbage results.  Basically like this:

```python
   def cmd(self):
        if self.view.is_dirty() or not self.view.file_name():
            if self.settings.get("lint_mode") in ("load_save", "save"):
                window = self.view.window()
                if window:
                    window.status_message("please save first")
                raise TransientError("Abort.  Buffer is dirty.")
            return ('phpmd', '${temp_file}', 'text')
            
        self.tempfile_suffix = '-'
        return ('phpmd', '${file_on_disk}', 'text')
```

And I think this code makes it clear why this is fishy, subtle.  Why don't we use `$temp_file` and have to save first?  What has that to do with the "mode"?   And then you replace the conditional, like so

```python
   def cmd(self):
        if self.view.is_dirty() or not self.view.file_name():
            if self.settings["only_lint_saved_files"]:
                window = self.view.window()
                if window:
                    window.status_message("please save first")
                raise TransientError("Abort.  Buffer is dirty.")
            return ('phpmd', '${temp_file}', 'text')
            
        self.tempfile_suffix = '-'
        return ('phpmd', '${file_on_disk}', 'text')
```

and then it is at least clear that this is an arbitrary decision, a user choice.  

Does that makes sense?

Reply:
> No, real-file linters only run on saved files. If you hit "lint this view" they (silently) skip. 

Ah, right, forgot about that. That changes things..

I like your suggestions, I’ll play with that a little bit tomorrow, see what that feels 
Ieke in practice.

Reply:
Alright, I pushed something that works. 

I like using the "mode" as the behaviour switch here. A different new setting, like "only_lint_saved_files" would interact weirdly with that mode setting: if you set the mode to "background" and "only_lint_saved_files" to true, what does that even mean? 

The only subtle side effect of using the mode as I'm doing here, is that you no longer get to use the manual trigger on dirty files. Edge cases that we can message about. Setting to "save" or "load_save" mode already implies that you intend to run on real files anyway.

Eh, all options have downsides, we can bike-shed all day here 🤷🏻

Reply:
That's not bike-shedding, you do a breaking a change here and restrict the possible usages, for example that you want to lint automatically on save but also at your-will, per shortcut.  

"only_lint_saved_files" is maybe misleading.  Technically you switch the operation mode of the linter.  We have stdin, temp file and real file modes.   And these are not switchable by the user as they would typically involve also changing the `cmd`, but here you want to opt into real-file mode.  Is "switch_to_real_file_mode: true" better?  I don't think so because why should we introduce the modes here in such a way?  "support_baseline_files: true" says what the switch is supposed to work-around but hides the side effect of loosing the while-you-type feature, hence comes as a surprise. 

> if you set the mode to "background" and "only_lint_saved_files" to true, what does that even mean?

The default for all linters is "background" and the real-file linters "downgrade" automatically.  That is maybe confusing but also already a present conception.

Reply:
Alright, color me convinced, you make good points all around. Like I said, I'm rusty on the subject and don't have all the modes and settings clear in my head. So, thanks 👍🏻 

I would say the new setting should have a name that closely represents _what_ it does, not _why_ (there can be many more, sometimes subtle "why's", there is only one "what"). And the "what" also indicates the side effect of no longer running on dirty buffers as you type. We can explain _why_ you'd want to use that setting in the README.

I think `"real_file_mode": true/false` is clear and simple.

Reply:
`real_file_mode` sounds really preferable.

It just leads to readable code, like I have a snippet some comments above, that would read now::

```
    def cmd(self):
        target = '$file_on_disk' if self.settings["real_file_mode"] else '$temp_file'
        return ('phpmd', target, 'text')

```

How can this be better, really?

Reply:
So, pretty 🥹 Python can be really poetic ❤️